### DISCUSSION TITLE
TIDAL prompt for oAuth -- it works

### DISCUSSION THREAD

**Discussion by guprobr** [strawberrymusicplayer/strawberry#1461] [Show and tell]
I never ever managed to login to Tidal via Strawberry. Until now. I had to fiddle with a lot of stuff and write a new function that prompts for the URL, so the AuthorizationUrlReceived gets called. Take a look guys, its working!
![Screenshot from 2024-06-12 22-01-09](https://github.com/strawberrymusicplayer/strawberry/assets/40776097/d7f03ae0-0572-4674-b65b-6298ffbf686e)
![Screenshot from 2024-06-12 22-01-29](https://github.com/strawberrymusicplayer/strawberry/assets/40776097/93946994-0e55-4f30-a638-a3fc168998dd)

I had to: 

* change the Scopes parameters, put "playback" instead
* change the variable that configures redirectURI to https://www.tidal.com
* create an APP in Tidal and set redirect url to https://www.tidal.com

The prompt has a text input where the user pastes the URL received on browser after all this conditions meet and after he clicks USe Oauth with proper client id from developers.tidal.com app

Login worked ! Nice huh?

**Comment by guprobr** [comment]
```
void TidalService::StartAuthorization(const QString &client_id) {

  client_id_ = client_id;
  code_verifier_ = Utilities::CryptographicRandomString(44);
  code_challenge_ = QString::fromLatin1(QCryptographicHash::hash(code_verifier_.toUtf8(), QCryptographicHash::Sha256).toBase64(QByteArray::Base64UrlEncoding));

  if (code_challenge_.lastIndexOf(QLatin1Char('=')) == code_challenge_.length() - 1) {
    code_challenge_.chop(1);
  }

  const ParamList params = ParamList() << Param(QStringLiteral("response_type"), QStringLiteral("code"))
                                       << Param(QStringLiteral("code_challenge"), code_challenge_)
                                       << Param(QStringLiteral("code_challenge_method"), QStringLiteral("S256"))
                                       << Param(QStringLiteral("redirect_uri"), QLatin1String(kOAuthRedirectUrl))
                                       << Param(QStringLiteral("client_id"), client_id_)
                                       << Param(QStringLiteral("scope"), QStringLiteral("playback"));

  QUrlQuery url_query;
  for (const Param &param : params) {
    url_query.addQueryItem(QString::fromLatin1(QUrl::toPercentEncoding(param.first)), QString::fromLatin1(QUrl::toPercentEncoding(param.second)));
  }

  QUrl url = QUrl(QString::fromLatin1(kOAuthUrl));
  url.setQuery(url_query);
  QDesktopServices::openUrl(url);

  // Prompt user and allow them to trigger AuthorizationUrlReceived manually
  // This could be a button click, menu option, or a message box
  TidalService::showAuthorizationPrompt();

}

#include <QtWidgets>

void TidalService::showAuthorizationPrompt() {
  // Create a dialog window
  QDialog dialog;
  dialog.setWindowTitle(QStringLiteral("Tidal Authorization"));

  // Create a layout for the dialog
  QVBoxLayout* layout = new QVBoxLayout(&dialog);

  // Label with instructions
  QLabel* instructionsLabel = new QLabel(
      QStringLiteral("**To authorize access to your Tidal account for [Strawberry Music Player], please follow these steps:**"));
  instructionsLabel->setWordWrap(true);
  layout->addWidget(instructionsLabel);

  // List of instructions
  QStringList instructions;
  instructions << QStringLiteral("1. Click the \"Authorize with Tidal\" button (or provide the appropriate button/menu option).");
  instructions << QStringLiteral("2. This will open a web page in your external browser where you can log in to your Tidal account.");
  instructions << QStringLiteral("**Important:** Once you've successfully logged in and authorized access, **do not close the browser window**.");
  instructions << QStringLiteral("3. Copy the entire URL from the address bar of your browser after the authorization is complete.");
  instructions << QStringLiteral("4. Paste the copied URL into the field below and click \"Continue\".");

  // List widget for instructions
  QListWidget* instructionsList = new QListWidget(&dialog);
  instructionsList->addItem(instructions.join(QStringLiteral("\n")));
  instructionsList->setEditTriggers(QAbstractItemView::NoEditTriggers);
  layout->addWidget(instructionsList);

  // Input field for URL
  QHBoxLayout* urlInputLayout = new QHBoxLayout;
  QLabel* urlLabel = new QLabel(QStringLiteral("Authorization URL:"));
  urlLabel->setAlignment(Qt::AlignRight | Qt::AlignVCenter);
  QLineEdit* urlLineEdit = new QLineEdit(&dialog);
  urlInputLayout->addWidget(urlLabel);
  urlInputLayout->addWidget(urlLineEdit);
  layout->addLayout(urlInputLayout);

  // Button to continue
  QPushButton* continueButton = new QPushButton(QStringLiteral("Continue"));
  connect(continueButton, &QPushButton::clicked, &dialog, &QDialog::accept);
  layout->addWidget(continueButton);

  // Button to cancel
  QPushButton* cancelButton = new QPushButton(QStringLiteral("Cancel"));
  connect(cancelButton, &QPushButton::clicked, &dialog, &QDialog::reject);
  layout->addWidget(cancelButton);

  // Show the dialog and wait for user input
  if (dialog.exec() == QDialog::Accepted) {
    QString url = urlLineEdit->text().trimmed();
    // Process the pasted URL (call AuthorizationUrlReceived)
    AuthorizationUrlReceived(QUrl(url));
  }
}

void TidalService::AuthorizationUrlReceived(const QUrl &url) {

  qLog(Debug) << "Tidal: Authorization URL Received" << url;

  QUrlQuery url_query(url);

```

**Reply by jonaski** [reply]
I don't know why it didn't work for you, I know many users use it so it should work if you have a compatible client ID.

**Reply by guprobr** [reply]
It gives an error on the oAuth screen. It does not accept the redirect URL nor the scopes

**Reply by guprobr** [reply]
my account is very new, maybe someething is changing?

**Comment by guprobr** [comment]
correction: missing r_usr and w_usr  scopes maybe

edit: this api.tidalhifi.com is not an official endpoint, right? There is discrepancy in the scopes! 

AUTHORIZATION in tidal official API

_Available scopes_
**user.read**
Read access to a user's account information, such as country and email address.

**collection.read**
Read access to a user's "My Collection".

**collection.write**
Write access to a user's "My Collection".

**playlists.read**
Read access to a user's playlists.

**playlists.write**
Write access to a user's playlists.

**recommendations.read**
Read access to a user’s personal recommendations.

**playback**
Required to play media content and control playback.

**Comment by guprobr** [comment]
hm these scopes r_usr and w_usr generates error when authenticating
but if I put NO SCOPE parameter at all, I get logged in and now the problem is this:

![Screenshot from 2024-06-12 22-12-59](https://github.com/strawberrymusicplayer/strawberry/assets/40776097/967c9782-7429-4360-bffa-40bb11a4fb2a)

edit: the countryCode is not returned by openapi.tidal.com
The code expects a "user" object in the JSON, containing some stuff. This does not exist in the responde of the real API.

**Comment by guprobr** [comment]
actually I turned to the debug messages, this probably indicates its very far from working still :( :( :( 
Its not even populating the JSON, sry this was a false positive success. :(

```
23:24:44.443 DEBUG TidalService:380                 Tidal: Authorization URL Received QUrl("https://tidal.com/?code=***CODE****&state=na")
23:24:44.729 DEBUG TidalService:572                 Tidal: Login successful user id 0 session id "" country code ""
23:25:10.067 ERROR TidalRequest:1391                Tidal: "countryCode parameter missing (400) (1002)"
```
edit: I forced my countryCode as a QStringLiteral and got to populate the configs with the **user_id** returned by the official api JSON (with proper name).
Now I realized the rules for the queries differ from api.tidalhifi.com and the official API. wow this will be a true adventure, to fix Tidal in Strawberry.. the openapi.tidal.com returns 203, 299 errors, very bizarre, so i'm stuck now on this. I think it requires some headers, the bearer auth and 
Content Type

application/vnd.tidal.v1+json


But i dunno how to fixx this, because the class is protected, i cant edit the headers from the Tidal class of our code... shucks.

**Reply by jonaski** [reply]
api.tidalhifi.com is the official closed API. The openapi.tidal.com is a new "open" API, but it does not provide any endpoint to actual get a stream URL, so if we should switch we need to find the endpoint, and perhaps also perform decryption of the stream.

**Reply by guprobr** [reply]
I see. I saw your post on MPD discussions, from 2019
Do you think it would be better to remove support for Tidal? or its me that cannot login and browse LoL
apparently, then, its the Scopes discrepancy the most trouble for oAuth to work AND still use the old api

**Comment by guprobr** [comment]
Well, in sum, the oauth login DID work: it returns properly the user_id
but there's those other issues.

I just saw the post on the MPD discussion board................. I'm starting to understand that perphaps it would be better to remove support for Tidal, or wait even more for a better public API culture by part of the Tidal company :(
looks like as if they don't even want apps that are compatible with their services.

I need a break, gonna rest and perphaps take a look at tidal-dl source code, stuff like that :(

**Reply by jonaski** [reply]
I removed Tidal a few years ago and got lot's of complaints from users who were using and were happy with it, so I ended up adding it back. But I know there are users who struggle to get it working, and I can't give much support since it's an unofficial solution and we can't get official documentation and support from Tidal. I even keep a Tidal subscription active which I don't actually use just for the purpose of testing and maintaining support for Tidal.
But it's nice that finally someone else is interested in this code.

**Comment by guprobr** [comment]
Redirect URLs in production apps must conform to the following rules: no query parameters, **only HTTPs schema, no localhost, no Private IP** ranges

thats one issue, thats why when I changed the redirectURL it started to work. The other issue are the scopes, r_usr and w_usr does not exist in the new loop.

**Reply by jonaski** [reply]
When using tidalhifi or opentidal API?

**Comment by guprobr** [comment]
When using tidalhifi, the new scopes bugs the search interface with error missing scope r_usr
If I try oAuth with the default cfg, it gives an error because the schema is not https for the redirectURL and the scopes does not exist. If I put user.read and playback, it works, authenticates, but then when in the search interface, the tidalhifi complains.

also the countrycode does not return by official oAuth. and user_id is the right name too, not userId

does not need to say, if I chsange to openapi.tidal.com, it gets even worse. Gives bizarre return codes like 203 and 299 lol

**Reply by Terzai56** [reply]
I have the exact same issue, I need user.read and playback for Auth. But when I query to find an album, I need r_usr scope.
Not sure what to do anymore ...
